conjure-undertow-processor/src/main/java/com/palantir/conjure/java/undertow/processor/data/ParamTypesResolver.java

```diff
@@ -98,23 +104,73 @@ public ParamTypesResolver(ResolverContext context) {
             return Optional.of(ParameterTypes.header(
                     annotationReflector.getAnnotationValue(String.class),
                     deserializerName,
-                    Instantiables.instantiate(annotationReflector.getAnnotationValue("decoder", TypeMirror.class))));
+                    getCollectionParamDecoder(variableElement, annotationReflector)));
         } else if (annotationReflector.isAnnotation(Handle.PathParam.class)) {
             String javaParameterName = variableElement.getSimpleName().toString();
             String deserializerName = InstanceVariables.joinCamelCase(javaParameterName, "Deserializer");
             return Optional.of(ParameterTypes.path(
-                    javaParameterName,
-                    deserializerName,
-                    Instantiables.instantiate(annotationReflector.getAnnotationValue("decoder", TypeMirror.class))));
+                    javaParameterName, deserializerName, getParamDecoder(variableElement, annotationReflector)));
         } else if (annotationReflector.isAnnotation(Handle.QueryParam.class)) {
             String deserializerName = InstanceVariables.joinCamelCase(
                     variableElement.getSimpleName().toString(), "Deserializer");
             return Optional.of(ParameterTypes.query(
                     annotationReflector.getAnnotationValue(String.class),
                     deserializerName,
-                    Instantiables.instantiate(annotationReflector.getAnnotationValue("decoder", TypeMirror.class))));
+                    getCollectionParamDecoder(variableElement, annotationReflector)));
         }
 
         throw new SafeIllegalStateException("Not possible");
     }
+
+    // TODO(fwindheuser): Bit messy, clean up
+
+    private CodeBlock getParamDecoder(VariableElement variableElement, AnnotationReflector annotationReflector) {
+        // If the default marker interface is not used (overwritten by user), we want to use the user-provided decoder.
+        TypeMirror typeMirror = annotationReflector.getAnnotationValue("decoder", TypeMirror.class);
+        if (!context.isSameTypes(typeMirror, DefaultParamDecoder.class)) {
+            return Instantiables.instantiate(typeMirror);
+        }
+
+        TypeMirror variableType = variableElement.asType();
+        String decoderMethodName =
+                DefaultDecoderNames.getDefaultDecoderMethodName(variableType, ContainerType.NONE, ContainerType.NONE);
+
+        // TODO(fwindheuser): We should reference the "runtime" var through a javapoet reference ('$N').
```

Comment:
Unclear that it would be any better, unless we wanted to rename the variable later, no strong opinion from me :-)

Reply:
Yeah, we would actually need to funnel the reference down through multiple methods, so skipping on this for now.